dali/operators/image/crop/bbox_crop.h

```diff
@@ -15,223 +15,27 @@
-      (height_offset + scaled_height));
-  }
-
-  const std::pair<BoundingBoxes, std::vector<int>> DiscardBoundingBoxesByCentroid(
-      const Crop &crop, const BoundingBoxes &bounding_boxes,
-      const std::vector<int> &labels) const {
-    DALI_ENFORCE(bounding_boxes.size() == labels.size(),
-                 "Labels and bounding boxes should have the same length");
-    BoundingBoxes candidate_boxes;
-    candidate_boxes.reserve(bounding_boxes.size());
-
-    std::vector<int> candidate_labels;
-    candidate_labels.reserve(labels.size());
-
-    // Discard bboxes whose centroid is not in the cropped area
-    for (size_t i = 0; i < bounding_boxes.size(); i++) {
-      auto coord = bounding_boxes[i].AsLtrb();
-      const float x_center = 0.5f * (coord[2] - coord[0]) + coord[0];
-      const float y_center = 0.5f * (coord[3] - coord[1]) + coord[1];
-
-      if (crop.Contains(x_center, y_center)) {
-        candidate_boxes.push_back(bounding_boxes[i]);
-        candidate_labels.push_back(labels[i]);
-      }
-    }
-
-    return std::make_pair(candidate_boxes, candidate_labels);
-  }
-
-  const ProspectiveCrop FindProspectiveCrop(
-      const BoundingBoxes &bounding_boxes, const std::vector<int> &labels,
-      std::pair<float, bool> minimum_overlap,
-      int sample) {
-    if (!minimum_overlap.second)
-      return ProspectiveCrop(true, Crop::FromLtrb(0, 0, 1, 1), bounding_boxes, labels);
-
-  std::uniform_real_distribution<float> float_dis_(scaling_bounds_.min, scaling_bounds_.max);
-    for (int i = 0; i < num_attempts_; ++i) {
-      // Image is HWC
-      const auto candidate_width = SampleCandidateDimension(sample);
-      const auto candidate_height = SampleCandidateDimension(sample);
-
-      if (ValidAspectRatio(candidate_height, candidate_width)) {
-        const auto candidate_crop =
-            SamplePatch(candidate_height, candidate_width, sample);
-
-        if (ValidOverlap(candidate_crop, bounding_boxes, minimum_overlap.first)) {
-          BoundingBoxes candidate_boxes;
-          std::vector<int> candidate_labels;
-
-          std::tie(candidate_boxes, candidate_labels) =
-                DiscardBoundingBoxesByCentroid(candidate_crop, bounding_boxes, labels);
-
-          if (candidate_boxes.begin() != candidate_boxes.end()) {
-            const auto remapped_boxes = RemapBoxes(
-              candidate_crop, candidate_boxes, candidate_height, candidate_width);
-            return ProspectiveCrop(true, candidate_crop, remapped_boxes, candidate_labels);
-          }
-        }
-      }
-    }
-    return ProspectiveCrop();
-  }
-
-  // float - threshold for IoU, bool - whether to apply crop (false means no cropp)
-  std::vector<std::pair<float, bool>> sample_options_;
-  const Bounds scaling_bounds_;
-  const Bounds aspect_ratio_bounds_;
-  const bool ltrb_;
-  const int num_attempts_;
-
  private:
-  BatchRNG<std::mt19937> rngs_;
-  USE_OPERATOR_MEMBERS();
+  class Impl;
+  std::unique_ptr<Impl> impl_;
```

Comment:
If you have pImpl, you can just as well have many implementations with static ndim. It would avoid a great deal of run-time checks, validations and assertions.

Reply:
done